components/fxa-client/src/lib.rs

```diff
@@ -89,30 +83,78 @@ impl FirefoxAccount {
     ///
     /// This method constructs as new [`FirefoxAccount`] instance configured to connect
     /// the application to a user's account.
-    ///
-    /// # Arguments
-    ///
-    ///   - `content_url` - the URL of the Firefox Accounts server to use
-    ///       - For example, use `https://accounts.firefox.com` for the main
-    ///         Mozilla-hosted service.
-    ///   - `client_id` - the registered OAuth client id of the application.
-    ///   - `redirect_uri` - the registered OAuth redirect URI of the application.
-    ///   - `token_server_url_override`: optionally, URL for the user's Sync Tokenserver.
-    ///        - This can be used to support users who self-host their sync data.
-    ///          If `None` then it will default to the Mozilla-hosted Sync server.
-    pub fn new(
-        content_url: &str,
-        client_id: &str,
-        redirect_uri: &str,
-        token_server_url_override: &Option<String>,
-    ) -> FirefoxAccount {
+    pub fn new(config: FxaConfig) -> FirefoxAccount {
         FirefoxAccount {
-            internal: std::sync::Mutex::new(internal::FirefoxAccount::new(
-                content_url,
-                client_id,
-                redirect_uri,
-                token_server_url_override.as_deref(),
-            )),
+            internal: std::sync::Mutex::new(internal::FirefoxAccount::new(config)),
```

Comment:
I'm thinking that things that block like the oauth handler, API client, storage handler, etc. should live here, next to the `Mutex<internal::FirefoxAccount>`.  That would facilitate things like releasing the lock while the application performs the oauth flow.  This is why I made the CLI examples construct the public FirefoxAccount rather than the internal one.

Once those changes are made, then there's going to be a bigger distinction between `FirefoxAccount` and `internal::FirefoxAccount`.  What do others think about changing the name to reflect that?  Maybe calling it `FxaInternal` or `FxaInternalState`?

There's an alternative would be to keep `FirefoxAccount` as a thin wrapper around `internal::FirefoxAccount`, and moving a lot of functionality into the `State` struct.  I don't think I like that as much though.

Reply:
I think it will all be clearer once we start seeing some of those changes, but here are my initial non-blocking thoughts

- By switching `internal::FirefoxAccount` into `internal::FxaInternalState`, are we going to consolidate the states into one place (putting the caches, the `State`, and the HTTP backoff state into one place and wrapping that up in the mutex?). If so I like it, in a way, we are collapsing the two `FirefoxAccount`s and making the locking more granular and +1 for that!
- How will the other members (HTTP client for example) interact with the state that requires locking the mutex? Will that be the responsibility of the high-level `FirefoxAccount`? For example after an http request if we want to update the back off for an endpoint?

Reply:
My goal is to move some of the state out of `internal::FirefoxAccount`.  I'm thinking the public `FirefoxAccount` will have fields that look something like this:
  - `FxaInternalState`: holds the current `State` field, plus some new internal state, like the current state transition.
  - `ApiClient`: contains the HTTP functionality + the caches and rate limiter/backoff
  - `OAuthHandler`: The application-provided oauth handler
  - `StorageHandler`: application-provider storage handler
  - `EventHandler`, `...`: I think all other application-provided callback interfaces will probably also live here, but haven't thought all of it through.

Each of these will have its own locking, if they need it.  Let's take the `get_devices()` call as an example:
  - Call `internal_state.get_device_request()`.  That locks the state to figure out which URL, refresh token, etc. to use.  Once we know the request, we release the lock.
  - Call `api_client.execute_request(device_request)` to perform the request.
     - Part of performing the request means checking/saving to the cache which will lock a mutex internal to `ApiClient`
     - For auth-related requests it would also check the rate-limiter, which would currently require a mutex, but I think we can avoid that with a refactor
   - Finally return the result of the request, maybe after some conversion.

By having multiple top-level components, each with their own locking, we allow more concurrency.  In particular, we don't want to block on the user-interaction in OAuthhandler.  `FxaInternalState`, or whatever we end up calling it, will be able to access the internal state, but should not block on IO, network, user-interaction, etc.

Each of the components will be independent of each other so there's no chances of deadlocks.  Maybe they will need some lock-free communication channels though, for example `FxaInternalState` may need a way to signal to `ApiClient` that it should clear it's cache/reset it's rate limiter.  Maybe not though, I could see the top-level `FirefoxAccount` handling that.